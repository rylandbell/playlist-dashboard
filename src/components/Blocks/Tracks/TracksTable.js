import React, { Component } from 'react';
import TracksTableRow from './TracksTableRow';
import TracksTableHead from './TracksTableHead';

class TracksTable extends Component {
  constructor(props) {
    super(props);
    this.onMouseLeave = this.onMouseLeave.bind(this);
  }

  onMouseLeave() {
    this.props.handleMouseLeavesTracksTable();
  }

  render() {
    const filteredTracks = this.props.filteredTracks;

    return (
      <table
        className="tracks__table table table-condensed table-hover"
        onMouseLeave={this.onMouseLeave}
      >
        <TracksTableHead features={this.props.features} handleColumnHeadingClick={this.props.handleColumnHeadingClick} />
        <tbody>
          {filteredTracks.map((track, index) =>
            <TracksTableRow
              key={track.id + index}
              track={track}
              features={this.props.features}
              handleTrackRowHover={this.props.handleTrackRowHover}
            />
          )}
        </tbody>
      </table>
    );
  }
}

export default TracksTable;
