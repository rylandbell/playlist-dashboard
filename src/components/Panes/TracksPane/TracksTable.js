import React, { Component } from 'react';
import TracksTableRow from './TracksTableRow';

class SelectPlaylistsTable extends Component {
  constructor(props) {
    super(props);
    this.onMouseLeave = this.onMouseLeave.bind(this);
  }

  onMouseLeave() {
    this.props.handleMouseLeavesTracksTable();
  }

  render() {
    const allFilters = this.props.allFilters;
    const filteredTracks = this.props.filteredTracks;
    const filteredFeaturesData = this.props.filteredFeaturesData;

    return (
      <div className="tracks__scroll-window">
        <table className="tracks__table table table-condensed table-hover" onMouseLeave={this.onMouseLeave}>
          <thead>
            <tr>
              <th>Track</th>
              {allFilters.map(filter => <th className="text-center hidden-xs" key={filter.name} >{filter.shortName}</th>)}
            </tr>
          </thead>
          <tbody>
            {filteredTracks.map((track, index) => 
              <TracksTableRow key={track.track.id + track.added_at + index} track={track} featuresData={filteredFeaturesData[index]} {...this.props} />
            )}
          </tbody>
        </table>
      </div>
    );
  }
}

export default SelectPlaylistsTable;