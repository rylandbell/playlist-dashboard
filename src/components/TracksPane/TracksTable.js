import React, { Component } from 'react';
import TracksTableRow from './TracksTableRow';

class SelectPlaylistsTable extends Component {
  render() {
    const tracks = this.props.reduxState.selectedPlaylistTracks;
    const features = this.props.reduxState.audioFeatures;
    const filters = this.props.reduxState.filters;

    function filterByFeatures(track, index) {

      //don't do any filtering before features data loads
      if (!features || features.length < 1) {
        return true;
      }

      const filterBooleans = [];
      for (let key in filters) {
        if (filters.hasOwnProperty(key)) {
          let passesFilter = features[index][key] >= filters[key][0] && features[index][key] <= filters[key][1];
          filterBooleans.push(passesFilter);
        }
      }

      const passesFilters = filterBooleans.reduce(
        (accumulator, currentValue) => accumulator && currentValue,
        true
      )

      return passesFilters;
    }

    const filteredTracks = tracks.filter(filterByFeatures);
    const filteredFeatures = features.filter(filterByFeatures);

    return (
      <div className="tracks__table">
        <table className="table table-condensed">
          <thead>
            <tr>
              <th>Title</th>
              <th>Danceability</th>
              <th>Instrumentalness</th>
            </tr>
          </thead>
          <tbody>
            {filteredTracks.map((track, index) => 
              <TracksTableRow key={track.track.id + track.added_at + index} track={track.track} features={filteredFeatures[index]} {...this.props} />
            )}
          </tbody>
        </table>
      </div>
    );
  }
}

export default SelectPlaylistsTable;