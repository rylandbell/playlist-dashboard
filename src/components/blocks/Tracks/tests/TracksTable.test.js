import React from "react";
import ReactDOM from "react-dom";
import renderer from "react-test-renderer";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";

import TracksTable from "../TracksTable";
import { testFeaturesData, testFilteredTracks } from "./test-data";

describe("TracksTable component", () => {
  const props = {
  features: testFeaturesData,
  filteredTracks: testFilteredTracks,
  handleTrackRowHover: () => {},
  handleMouseLeavesTracksTable: () => {},
  handleColumnHeadingClick: () => {}
};

  it("renders without crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(<TracksTable {...props} />, div);
  });

  it("shallow render matches snapshot", () => {
    const tree = toJson(shallow(<TracksTable {...props} />));
    expect(tree).toMatchSnapshot();
  });
});
