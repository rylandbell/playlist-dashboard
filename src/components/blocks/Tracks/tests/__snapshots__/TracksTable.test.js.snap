// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TracksTable component shallow render matches snapshot 1`] = `
<table
  className="tracks__table table table-condensed table-hover"
  onMouseLeave={[Function]}
>
  <TracksTableHead
    features={
      Array [
        Object {
          "color": "hsl(194, 66%, 61%)",
          "currentValue": Array [
            0,
            1,
          ],
          "description": "Danceability describes how suitable a track is for dancing based on a combination of musical elements including tempo, rhythm stability, beat strength, and overall regularity. A value of 0.0 is least danceable and 1.0 is most danceable.",
          "dimColor": "hsl(194, 46%, 21%)",
          "displayName": "Danceability",
          "isActive": true,
          "isDim": true,
          "isGraphed": true,
          "max": 1,
          "min": 0,
          "name": "danceability",
          "shortName": "Dance",
          "showReferenceLine": true,
          "sortBy": false,
        },
        Object {
          "color": "hsl(2, 64%, 58%)",
          "currentValue": Array [
            0,
            1,
          ],
          "description": "Energy is a measure from 0.0 to 1.0 and represents a perceptual measure of intensity and activity. Typically, energetic tracks feel fast, loud, and noisy. For example, death metal has high energy, while a Bach prelude scores low on the scale. Perceptual features contributing to this attribute include dynamic range, perceived loudness, timbre, onset rate, and general entropy.",
          "dimColor": "hsl(2, 44%, 18%)",
          "displayName": "Energy",
          "isActive": true,
          "isDim": false,
          "isGraphed": false,
          "max": 1,
          "min": 0,
          "name": "energy",
          "shortName": "Energy",
          "showReferenceLine": false,
          "sortBy": false,
        },
        Object {
          "color": "hsl(120, 39%, 54%)",
          "currentValue": Array [
            0,
            1,
          ],
          "description": "A measure from 0.0 to 1.0 describing the musical positiveness conveyed by a track. Tracks with high valence sound more positive (e.g. happy, cheerful, euphoric), while tracks with low valence sound more negative (e.g. sad, depressed, angry).",
          "dimColor": "hsl(120, 19%, 14%)",
          "displayName": "Positivity",
          "isActive": true,
          "isDim": false,
          "isGraphed": false,
          "max": 1,
          "min": 0,
          "name": "valence",
          "shortName": "Pos",
          "showReferenceLine": false,
          "sortBy": false,
        },
        Object {
          "color": "hsl(24, 79%, 49%)",
          "currentValue": Array [
            0,
            1,
          ],
          "description": "The popularity of a track is a value between 0 and 1. It is calculated by algorithm and is based, in the most part, on the total number of plays the track has had and how recent those plays are.",
          "dimColor": "hsl(24, 59%, 19%)",
          "displayName": "Popularity",
          "isActive": true,
          "isDim": false,
          "isGraphed": false,
          "max": 1,
          "min": 0,
          "name": "popularity",
          "shortName": "Pop",
          "showReferenceLine": false,
          "sortBy": false,
        },
        Object {
          "color": "hsl(35, 84%, 62%)",
          "currentValue": Array [
            0,
            1,
          ],
          "description": "A confidence measure from 0.0 to 1.0 of whether the track is acoustic. 1.0 represents high confidence the track is acoustic.",
          "dimColor": "hsl(35, 64%, 17%)",
          "displayName": "Acousticness",
          "isActive": true,
          "isDim": false,
          "isGraphed": false,
          "max": 1,
          "min": 0,
          "name": "acousticness",
          "shortName": "Acous",
          "showReferenceLine": false,
          "sortBy": false,
        },
        Object {
          "color": "hsl(0, 0%, 92%)",
          "currentValue": Array [
            0,
            1,
          ],
          "description": "Predicts whether a track contains no vocals. \\"Ooh\\" and \\"aah\\" sounds are treated as instrumental in this context. Rap or spoken word tracks are clearly \\"vocal\\". The closer the instrumentalness value is to 1.0, the greater likelihood the track contains no vocal content. Values above 0.5 are intended to represent instrumental tracks, but confidence is higher as the value approaches 1.0.",
          "dimColor": "hsl(0, 0%, 22%)",
          "displayName": "Instrumentalness",
          "isActive": true,
          "isDim": false,
          "isGraphed": false,
          "max": 1,
          "min": 0,
          "name": "instrumentalness",
          "shortName": "Inst",
          "showReferenceLine": false,
          "sortBy": false,
        },
      ]
    }
    handleColumnHeadingClick={[Function]}
  />
  <tbody>
    <TracksTableRow
      features={
        Array [
          Object {
            "color": "hsl(194, 66%, 61%)",
            "currentValue": Array [
              0,
              1,
            ],
            "description": "Danceability describes how suitable a track is for dancing based on a combination of musical elements including tempo, rhythm stability, beat strength, and overall regularity. A value of 0.0 is least danceable and 1.0 is most danceable.",
            "dimColor": "hsl(194, 46%, 21%)",
            "displayName": "Danceability",
            "isActive": true,
            "isDim": true,
            "isGraphed": true,
            "max": 1,
            "min": 0,
            "name": "danceability",
            "shortName": "Dance",
            "showReferenceLine": true,
            "sortBy": false,
          },
          Object {
            "color": "hsl(2, 64%, 58%)",
            "currentValue": Array [
              0,
              1,
            ],
            "description": "Energy is a measure from 0.0 to 1.0 and represents a perceptual measure of intensity and activity. Typically, energetic tracks feel fast, loud, and noisy. For example, death metal has high energy, while a Bach prelude scores low on the scale. Perceptual features contributing to this attribute include dynamic range, perceived loudness, timbre, onset rate, and general entropy.",
            "dimColor": "hsl(2, 44%, 18%)",
            "displayName": "Energy",
            "isActive": true,
            "isDim": false,
            "isGraphed": false,
            "max": 1,
            "min": 0,
            "name": "energy",
            "shortName": "Energy",
            "showReferenceLine": false,
            "sortBy": false,
          },
          Object {
            "color": "hsl(120, 39%, 54%)",
            "currentValue": Array [
              0,
              1,
            ],
            "description": "A measure from 0.0 to 1.0 describing the musical positiveness conveyed by a track. Tracks with high valence sound more positive (e.g. happy, cheerful, euphoric), while tracks with low valence sound more negative (e.g. sad, depressed, angry).",
            "dimColor": "hsl(120, 19%, 14%)",
            "displayName": "Positivity",
            "isActive": true,
            "isDim": false,
            "isGraphed": false,
            "max": 1,
            "min": 0,
            "name": "valence",
            "shortName": "Pos",
            "showReferenceLine": false,
            "sortBy": false,
          },
          Object {
            "color": "hsl(24, 79%, 49%)",
            "currentValue": Array [
              0,
              1,
            ],
            "description": "The popularity of a track is a value between 0 and 1. It is calculated by algorithm and is based, in the most part, on the total number of plays the track has had and how recent those plays are.",
            "dimColor": "hsl(24, 59%, 19%)",
            "displayName": "Popularity",
            "isActive": true,
            "isDim": false,
            "isGraphed": false,
            "max": 1,
            "min": 0,
            "name": "popularity",
            "shortName": "Pop",
            "showReferenceLine": false,
            "sortBy": false,
          },
          Object {
            "color": "hsl(35, 84%, 62%)",
            "currentValue": Array [
              0,
              1,
            ],
            "description": "A confidence measure from 0.0 to 1.0 of whether the track is acoustic. 1.0 represents high confidence the track is acoustic.",
            "dimColor": "hsl(35, 64%, 17%)",
            "displayName": "Acousticness",
            "isActive": true,
            "isDim": false,
            "isGraphed": false,
            "max": 1,
            "min": 0,
            "name": "acousticness",
            "shortName": "Acous",
            "showReferenceLine": false,
            "sortBy": false,
          },
          Object {
            "color": "hsl(0, 0%, 92%)",
            "currentValue": Array [
              0,
              1,
            ],
            "description": "Predicts whether a track contains no vocals. \\"Ooh\\" and \\"aah\\" sounds are treated as instrumental in this context. Rap or spoken word tracks are clearly \\"vocal\\". The closer the instrumentalness value is to 1.0, the greater likelihood the track contains no vocal content. Values above 0.5 are intended to represent instrumental tracks, but confidence is higher as the value approaches 1.0.",
            "dimColor": "hsl(0, 0%, 22%)",
            "displayName": "Instrumentalness",
            "isActive": true,
            "isDim": false,
            "isGraphed": false,
            "max": 1,
            "min": 0,
            "name": "instrumentalness",
            "shortName": "Inst",
            "showReferenceLine": false,
            "sortBy": false,
          },
        ]
      }
      handleTrackRowHover={[Function]}
      track={
        Object {
          "acousticness": 0.575,
          "artist": "KMD",
          "danceability": 0.881,
          "duration_ms": 241626,
          "energy": 0.836,
          "id": "6UUY9L3jGnK507sKWK20s3",
          "instrumentalness": 0,
          "liveness": 0.105,
          "name": "Peachfuzz",
          "popularity": 0.35,
          "uri": "spotify:track:6UUY9L3jGnK507sKWK20s3",
          "valence": 0.461,
        }
      }
    />
    <TracksTableRow
      features={
        Array [
          Object {
            "color": "hsl(194, 66%, 61%)",
            "currentValue": Array [
              0,
              1,
            ],
            "description": "Danceability describes how suitable a track is for dancing based on a combination of musical elements including tempo, rhythm stability, beat strength, and overall regularity. A value of 0.0 is least danceable and 1.0 is most danceable.",
            "dimColor": "hsl(194, 46%, 21%)",
            "displayName": "Danceability",
            "isActive": true,
            "isDim": true,
            "isGraphed": true,
            "max": 1,
            "min": 0,
            "name": "danceability",
            "shortName": "Dance",
            "showReferenceLine": true,
            "sortBy": false,
          },
          Object {
            "color": "hsl(2, 64%, 58%)",
            "currentValue": Array [
              0,
              1,
            ],
            "description": "Energy is a measure from 0.0 to 1.0 and represents a perceptual measure of intensity and activity. Typically, energetic tracks feel fast, loud, and noisy. For example, death metal has high energy, while a Bach prelude scores low on the scale. Perceptual features contributing to this attribute include dynamic range, perceived loudness, timbre, onset rate, and general entropy.",
            "dimColor": "hsl(2, 44%, 18%)",
            "displayName": "Energy",
            "isActive": true,
            "isDim": false,
            "isGraphed": false,
            "max": 1,
            "min": 0,
            "name": "energy",
            "shortName": "Energy",
            "showReferenceLine": false,
            "sortBy": false,
          },
          Object {
            "color": "hsl(120, 39%, 54%)",
            "currentValue": Array [
              0,
              1,
            ],
            "description": "A measure from 0.0 to 1.0 describing the musical positiveness conveyed by a track. Tracks with high valence sound more positive (e.g. happy, cheerful, euphoric), while tracks with low valence sound more negative (e.g. sad, depressed, angry).",
            "dimColor": "hsl(120, 19%, 14%)",
            "displayName": "Positivity",
            "isActive": true,
            "isDim": false,
            "isGraphed": false,
            "max": 1,
            "min": 0,
            "name": "valence",
            "shortName": "Pos",
            "showReferenceLine": false,
            "sortBy": false,
          },
          Object {
            "color": "hsl(24, 79%, 49%)",
            "currentValue": Array [
              0,
              1,
            ],
            "description": "The popularity of a track is a value between 0 and 1. It is calculated by algorithm and is based, in the most part, on the total number of plays the track has had and how recent those plays are.",
            "dimColor": "hsl(24, 59%, 19%)",
            "displayName": "Popularity",
            "isActive": true,
            "isDim": false,
            "isGraphed": false,
            "max": 1,
            "min": 0,
            "name": "popularity",
            "shortName": "Pop",
            "showReferenceLine": false,
            "sortBy": false,
          },
          Object {
            "color": "hsl(35, 84%, 62%)",
            "currentValue": Array [
              0,
              1,
            ],
            "description": "A confidence measure from 0.0 to 1.0 of whether the track is acoustic. 1.0 represents high confidence the track is acoustic.",
            "dimColor": "hsl(35, 64%, 17%)",
            "displayName": "Acousticness",
            "isActive": true,
            "isDim": false,
            "isGraphed": false,
            "max": 1,
            "min": 0,
            "name": "acousticness",
            "shortName": "Acous",
            "showReferenceLine": false,
            "sortBy": false,
          },
          Object {
            "color": "hsl(0, 0%, 92%)",
            "currentValue": Array [
              0,
              1,
            ],
            "description": "Predicts whether a track contains no vocals. \\"Ooh\\" and \\"aah\\" sounds are treated as instrumental in this context. Rap or spoken word tracks are clearly \\"vocal\\". The closer the instrumentalness value is to 1.0, the greater likelihood the track contains no vocal content. Values above 0.5 are intended to represent instrumental tracks, but confidence is higher as the value approaches 1.0.",
            "dimColor": "hsl(0, 0%, 22%)",
            "displayName": "Instrumentalness",
            "isActive": true,
            "isDim": false,
            "isGraphed": false,
            "max": 1,
            "min": 0,
            "name": "instrumentalness",
            "shortName": "Inst",
            "showReferenceLine": false,
            "sortBy": false,
          },
        ]
      }
      handleTrackRowHover={[Function]}
      track={
        Object {
          "acousticness": 0.046,
          "artist": "Tidus",
          "danceability": 0.939,
          "duration_ms": 192000,
          "energy": 0.653,
          "id": "0tJWc3RXOuli1FjEgzpCn7",
          "instrumentalness": 0.277,
          "liveness": 0.0336,
          "name": "Delancey",
          "popularity": 0.41,
          "uri": "spotify:track:0tJWc3RXOuli1FjEgzpCn7",
          "valence": 0.81,
        }
      }
    />
  </tbody>
</table>
`;
